File: app/src/main/java/de/westnordost/streetcomplete/data/DbModule.java
package de.westnordost.streetcomplete.data;

import android.content.Context;
import android.content.SharedPreferences;
import android.database.sqlite.SQLiteOpenHelper;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import de.westnordost.streetcomplete.ApplicationConstants;
import de.westnordost.streetcomplete.data.visiblequests.QuestTypeOrderList;
import de.westnordost.streetcomplete.data.visiblequests.VisibleQuestTypeDao;
import de.westnordost.streetcomplete.util.KryoSerializer;
import de.westnordost.streetcomplete.util.Serializer;

@Module
public class DbModule
{
	@Provides @Singleton public static SQLiteOpenHelper sqliteOpenHelper(Context ctx)
	{
		return sqliteOpenHelper(ctx, ApplicationConstants.DATABASE_NAME);
	}

	public static SQLiteOpenHelper sqliteOpenHelper(Context ctx, String databaseName)
	{
		return new StreetCompleteSQLiteOpenHelper(ctx, databaseName);
	}

	@Provides @Singleton public static Serializer serializer()
	{
		return new KryoSerializer();
	}

	@Provides @Singleton public static VisibleQuestTypeDao visibleQuestTypeDao(
			SQLiteOpenHelper dbHelper)
	{
		return new VisibleQuestTypeDao(dbHelper);
	}

	@Provides @Singleton public static QuestTypeOrderList questTypeOrderDao(
			SharedPreferences prefs, QuestTypeRegistry questTypeRegistry)
	{
		return new QuestTypeOrderList(prefs, questTypeRegistry);
	}
}

File: app/src/androidMain/kotlin/de/westnordost/streetcomplete/data/DbModule.kt
package de.westnordost.streetcomplete.data

import android.content.Context
import androidx.sqlite.driver.bundled.BundledSQLiteDriver
import de.westnordost.streetcomplete.ApplicationConstants
import org.koin.dsl.module

val dbModule = module {
    single<Database> {
        val databaseFilePath = get<Context>().getDatabasePath(ApplicationConstants.DATABASE_NAME).path
        val databaseConnection = BundledSQLiteDriver().open(databaseFilePath)
        StreetCompleteDatabase(databaseConnection)
    }
}
